Oauth2 device flow 141

miltonbo opened the pull request:
@yuriyz Please, would you review this PR, it's related to the issue: #141 

I couldn't run Jenkins tests because that environment is not running correctly, also let me know if I have to restore that env. Like always any comment is welcome.

yuriyz commented on Server/src/main/java/org/gluu/oxauth/token/ws/rs/TokenRestWebServiceImpl.java:
```diff
@@ -490,6 +495,103 @@ public Response requestAccessToken(String grantType, String code,
+                cacheData.setLastAccessControl(currentTime);
+                deviceAuthorizationService.saveInCache(cacheData, true, true);
+
+                if (cacheData.getStatus() == DeviceAuthorizationStatus.PENDING) {
```
Minor : All validation checks can be written in form of WebApplicationException and be linear instead of if/else with another inner if/else. It's not always possible to get it linear but here it seems to be possible.

yuriyz commented on Server/src/main/java/org/gluu/oxauth/token/ws/rs/TokenRestWebServiceImpl.java:
```diff
@@ -490,6 +495,103 @@ public Response requestAccessToken(String grantType, String code,
+                        null, deviceCodeGrant, false, null);
+
+                deviceCodeGrant.setTokensDelivered(true);
+                deviceCodeGrant.save();
```
Once tokens are issued deviceCode has to be removed similar to AuthorizationCode.

yuriyz commented on Server/src/main/resources/oxauth_de.properties:
```diff
@@ -195,3 +195,15 @@ password.validation.invalid = The password provide is not strong enough
 
 # CIBA
 ciba.bindingMessage=Binding Message
+
```
What is the sense of pasting english into oxauth_de file? Resource bundle automatically falls back to oxauth.properties if key is not found. Same for other localionzation files.

miltonbo commented on Server/src/main/java/org/gluu/oxauth/authorize/ws/rs/AuthorizeRestWebServiceImpl.java:
```diff
@@ -142,13 +145,13 @@ public Response requestAuthorizationGet(
                 prompt, maxAge, uiLocales, idTokenHint, loginHint, acrValues, amrValues, request, requestUri,
                 requestSessionId, sessionId, HttpMethod.GET, originHeaders, codeChallenge, codeChallengeMethod,
-                customResponseHeaders, claims, authReqId, httpRequest, httpResponse, securityContext);
+                customResponseHeaders, claims, authReqId, userCode, httpRequest, httpResponse, securityContext);
```
Because authorization is done also using this and authorize page, during the whole process the identifier for device authz is userCode, it could be expired and also it allows get information from cache when authz is granted or denied.

miltonbo commented on Server/src/main/java/org/gluu/oxauth/authorize/ws/rs/DeviceAuthorizationAction.java:
```diff
@@ -0,0 +1,313 @@
+    public void init() {
+        lastAttempt = System.currentTimeMillis();
+        attempts = 0;
+        code = sessionId = state = sessionState = error = errorDescription = userCode = userCode1 = userCode2 = null;
```
Yeah, but this is also used when user wants to start again, there is a button in the UI that uses it, therefore the page state should be reset.

miltonbo replied:
This method was removed, since session scope is not longer available and we are using RequestScope.

miltonbo commented on Server/src/main/java/org/gluu/oxauth/service/DeviceAuthorizationService.java:
```diff
@@ -0,0 +1,149 @@
+            cacheService.put(data.getExpiresIn(), data.getDeviceCode(), data);
+        }
+        if (saveUserCode) {
+            cacheService.put(data.getExpiresIn(), data.getUserCode(), data);
```
Yes, also I thought that. I think it's enough for now, but should be improved.

miltonbo commented on Server/src/main/java/org/gluu/oxauth/authorize/ws/rs/DeviceAuthorizationAction.java:
```diff
@@ -0,0 +1,313 @@
+ * Action used to process all requests related to device authorization.
+ */
+@Named
+@SessionScoped
```
Ok, well, I changed it to use SessionId, only saved required data during session.

miltonbo commented:
@yuriyz I followed most of your suggestions, please review my comments and my changes, I think the bigest one was related to SessionScope.

Also I run Jenkins, all successful: https://ox.gluu.org/jenkins/job/oxAuth_master_LDAP/1112

About tests, I've been working on those also, however this PR is focused on functionalities and also it has too many changes. After approval I'll be working on those tests, configuration, docs, oxAuth-RP and oxTrust.

miltonbo commented:
@yuriyz After Jenkins build, all tests passed, including new ones that I implemented for device authorization flow. Please take a look to my changes and let me know whatever could be improved/changed.

https://ox.gluu.org/jenkins/job/oxAuth_master_LDAP/1156

Also I reverted my config in Json properties.

yuriyz commented on Client/src/main/java/org/gluu/oxauth/client/DeviceAuthzClient.java:
```diff
@@ -0,0 +1,107 @@
+        } catch (Exception e) {
+            e.printStackTrace();
+            return null;
+        }
```
Connection is not released. Please release it as it's done in other client classes.

yuriyz commented on Client/src/main/java/org/gluu/oxauth/client/DeviceAuthzClient.java:
```diff
@@ -0,0 +1,107 @@
+
+            return getResponse();
+        } catch (Exception e) {
+            e.printStackTrace();
```
Log exception in logger.

yuriyz commented on Client/src/main/java/org/gluu/oxauth/client/DeviceAuthzClient.java:
```diff
@@ -0,0 +1,107 @@
+            clientRequest.setHttpMethod(getHttpMethod());
+            clientRequest.header("Content-Type", request.getContentType());
+            if (request.getAuthenticationMethod() == AuthenticationMethod.CLIENT_SECRET_BASIC && request.hasCredentials()) {
+                clientRequest.header("Authorization", "Basic " + request.getEncodedCredentials());
```
I guess we should have other client authentication types support here. See `TokenClient` class.

yuriyz commented on Client/src/main/java/org/gluu/oxauth/client/DeviceAuthzRequest.java:
```diff
@@ -0,0 +1,112 @@
+                parameters.put(key, getCustomParameters().get(key));
+            }
+        } catch (JSONException e) {
+            e.printStackTrace();
```
Log in logger, not just print in console.

yuriyz commented on Client/src/main/java/org/gluu/oxauth/client/DeviceAuthzRequest.java:
```diff
@@ -0,0 +1,112 @@
+                queryStringBuilder.append(key).append("=").append(getCustomParameters().get(key));
+            }
+        } catch (UnsupportedEncodingException | JSONException e) {
+            e.printStackTrace();
```
Logger.

yuriyz commented on Client/src/test/java/org/gluu/oxauth/ws/rs/deviceauthz/DeviceAuthzFlowHttpTest.java:
```diff
@@ -0,0 +1,605 @@
+        tokenRequest.setDeviceCode(deviceCode);
+
+        TokenClient tokenClient1 = newTokenClient(tokenRequest);
+        tokenClient1.setRequest(tokenRequest);
```
This line is redundant, request already passed in `newTokenClient(tokenRequest);`

yuriyz commented on Client/src/test/java/org/gluu/oxauth/ws/rs/deviceauthz/DeviceAuthzFlowHttpTest.java:
```diff
@@ -0,0 +1,605 @@
+        assertNull(authorizationResponse.getErrorType());
+    }
+
+    private void processDeviceAuthzPutUserCodeAndPressContinue(String userCode, WebDriver currentDriver, boolean complete) {
```
Not critical but I guess we can organize selenium interaction in separate object which fits better to OOP. E.g. as in `SelectPage`.

yuriyz commented on Server/src/main/java/org/gluu/oxauth/service/AuthorizeService.java:
```diff
@@ -298,4 +297,17 @@ private boolean invalidateSessionCookies() {
+        DeviceAuthorizationCacheControl cacheData = deviceAuthorizationService.getDeviceAuthzByUserCode(userCode);
+
+        if (cacheData != null) {
+            if (cacheData.getStatus() == DeviceAuthorizationStatus.PENDING) {
```
Maybe && without 2 ifs ?

yuriyz commented on Server/src/main/java/org/gluu/oxauth/authorize/ws/rs/AuthorizeAction.java:
```diff
@@ -364,7 +346,8 @@ public void checkPermissionGranted() throws IOException {
         }
 
-        if (StringUtils.isBlank(redirectionUriService.validateRedirectionUri(clientId, redirectUri))) {
+        String userCode = session.getSessionAttributes().get(AuthorizeRequestParam.USER_CODE);
```
SESSION_USER_CODE

yuriyz commented on Server/src/main/java/org/gluu/oxauth/service/AuthorizeService.java:
```diff
@@ -298,4 +297,17 @@ private boolean invalidateSessionCookies() {
     }
+
+    private void processDeviceAuthDeniedResponse(Map<String, String> sessionAttribute) {
+        String userCode = sessionAttribute.get(AuthorizeRequestParam.USER_CODE);
```
SESSION_USER_CODE

yuriyz commented on Model/src/main/java/org/gluu/oxauth/model/authorize/AuthorizeRequestParam.java:
```diff
@@ -38,6 +38,7 @@
     String CODE_CHALLENGE_METHOD = "code_challenge_method";
     String CUSTOM_RESPONSE_HEADERS = "custom_response_headers";
     String AUTH_REQ_ID = "auth_req_id";
+    String USER_CODE = "user_code";
```
We shouldn't have it anymore here. It's not part of request. We should have internal SESSION_USER_CODE.

miltonbo commented:
@yuriyz I did all required changes.

Jenkins build:
https://ox.gluu.org/jenkins/job/oxAuth_master_LDAP/1159

Also I removed device authz config.
